qdglue/metrics/grid_based_metrics.py

```diff
@@ -0,0 +1,290 @@
+import dataclasses
+from abc import ABC, abstractmethod
+from typing import Dict, Tuple
+
+import gymnasium
+import numpy as np
+
+from qdglue.types import Feature, Fitness
+
+
+@dataclasses.dataclass
+class DiscreteArchiveMetrics:
+    """Holds statistics about an archive.
+
+    Attributes:
+        coverage: Proportion of cells in the archive that have an elite - always
+            in the range $[0,1]$.
+        qd_score: QD score, i.e. sum of objective values of all elites in the
+            archive. All the fitnesses are normalized to $[0,1]$ using the
+            fitness_bounds before computing the QD score.
+        max_fitness: Maximum objective value of the elites in the archive.
+        ccdf: Complementary Cumulative Distribution of Fitness from Vassiliades
+            et al. (2018)
+    """
+
+    coverage: float
+    qd_score: float
+    max_fitness: float
+    ccdf: np.ndarray
+
+
+class DiscreteArchiveMetricsCalculator(ABC):
+    def __init__(
+        self,
+        fitness_bounds: gymnasium.spaces.Box,
+        num_points_ccdf: int = 100,
+    ):
+        assert np.all(np.isfinite(fitness_bounds.low)) and np.all(
+            np.isfinite(fitness_bounds.high)
+        ), "Fitness bounds must be finite"
+        assert (
+            np.prod(fitness_bounds.shape) == 1
+        ), "Only scalar fitnesses are supported for now"
+
+        self.fitness_bounds = fitness_bounds
+        self.num_points_ccdf = num_points_ccdf
+
+    def get_metrics(
+        self,
+        fitnesses: Fitness,
+        features: Feature,
+    ) -> DiscreteArchiveMetrics:
+        self._check_validity_features(features)
+        self._check_validity_fitnesses(fitnesses)
+
+        # Extracting the best fitnesses per cell
+        # The empty cells do not appear in the best_fitnesses array
+        best_fitnesses = self._extract_best_fitnesses_per_cell(fitnesses, features)
+        best_fitnesses_normalized = (best_fitnesses - self.fitness_bounds.low) / (
+            self.fitness_bounds.high - self.fitness_bounds.low
+        )
+        # TODO(looka): support case where fitness_bounds.high == +inf, and just apply the offset in that case?
+
+        # Computing the metrics
+        total_cells = self.num_cells
+
+        coverage = np.size(best_fitnesses) / total_cells
+        max_fitness = np.max(best_fitnesses)
+        qd_score = np.sum(best_fitnesses_normalized).item()
+
+        # Computing the Complementary Cumulative Distribution of Fitness
+        ccdf = self._compute_ccdf(best_fitnesses)
+
+        return DiscreteArchiveMetrics(
+            coverage=coverage,
+            qd_score=qd_score,
+            max_fitness=max_fitness,
+            ccdf=ccdf,
+        )
+
+    def _compute_ccdf(
+        self,
+        best_fitnesses: Fitness,
+    ) -> np.ndarray:
+        fitnesses_intermediate = np.linspace(
```

Comment:
My feeling is that `np.histogram` will do what you're trying to do here, but I can't remember how. Can you take a look? I think it will reduce this function to roughly one function call.

Reply:
Nice! `np.histogram` with an inverted `cumsum` does the job! Thank you so much! ^^